src/ds/ptrwrap.h

```diff
@@ -38,67 +38,138 @@ namespace snmalloc
    * *E forms are "exported" and have had platform constraints applied.  That
    * means, for example, on CheriBSD, that they have had their VMMAP permission
    * stripped.
-   *
-   * Yes, I wish the start-of-comment characters were aligned below as well.
-   * I blame clang format.
    */
-  enum capptr_bounds
+
+  namespace capptr_bounds
   {
-    /*                Spatial  Notes                                      */
-    CBArena, /*        Arena                                              */
-    CBChunkD, /*       Arena   Chunk-bounded in debug; internal use only! */
-    CBChunk, /*        Chunk                                              */
-    CBChunkE, /*       Chunk   (+ platform constraints)                   */
-    CBAlloc, /*        Alloc                                              */
-    CBAllocE /*        Alloc   (+ platform constraints)                   */
+    /*
+     * Bounds dimensions are sorted so that < reflects authority.  For example,
+     * spatial::Alloc is less capable than spatial::Arena.
+     */
+    enum class spatial
+    {
+      Alloc,
+      Chunk,
+      ChunkD,
+      Arena
+    };
+    enum class platform
+    {
+      Exported,
+      High
+    };
+    enum class wild
+    {
+      Unchecked,
+      Checked
```

Comment:
We have to be careful about changing from `Unchecked` to `Checked` in particular, we could end up encoding a "time of check to time of use" attack into the API design?

Reply:
Yes, and this is more subtle than I'd like.  A `wild::Checked` pointer just means that we have validated, on CHERI, that the capability is tagged and that it points at its base (and nothing on non-CHERI).  It doesn't mean that it's within an allocation arena or corresponds to an allocated object.

In the final taxonomy, I expect pointers from the client to be
- `wild::Unchecked`
- `platform::Exported`
- `temporal::Monochrome`
- `spatial::Alloc`
- `zero::Arbitrary`
and then we will check (and change)...
- capability attributes (convert to `wild::Checked`)
- that the pointer points into an arena (via amplification, giving us a `wild::Checked`, `platform::High`, `temporal::Rainbow`, `spatial::Arena`, `zero::Arbitrary` analogue to the client pointer)
- that we have been told the right size (for size-parametric `dealloc`s; see `*_dealloc_unchecked` and `*_dealloc_checked_chunkmap`).
- that, in fact, it points to the start of an object (by determining the sizeclass and checking `is_start_of_object` or analogue; see `*_dealloc_checked_sizeclass`; at this point we can convert down to a `spatial::Alloc` pointer again)
- that the MTE colours aligned (using our `temporal::Rainbow` rights to atomically clock the 0th granule of the object from the provided colour to the next; see placeholders in `*_dealloc_start`; at this point, we can convert down to `temporal::Monochrome` and `zero::Zeroed` when we `bzero_with_color()` the memory)

Phew!  And then we can thread the thing on a `Remote` queue or hand it off to the backing {`Slab`, `Mediumslab`, `LargeAlloc`} for dealing with.